pkg/frontend/querymiddleware/results_cache.go

```diff
@@ -174,19 +174,22 @@ type shouldCacheFn func(r Request) bool
 var resultsCacheAlwaysEnabled = func(_ Request) bool { return true }
 
 // isRequestCachable says whether the request is eligible for caching.
-func isRequestCachable(req Request, maxCacheTime int64, cacheUnalignedRequests bool, logger log.Logger) bool {
+func isRequestCachable(req Request, maxCacheTime int64, cacheUnalignedRequests bool, logger log.Logger, notCachableReasonCount *prometheus.CounterVec) bool {
```

Comment:
[nit] From a design perspective, I would rather return `(cachable bool, reason string)` from this function and let the caller increase the metric when appropriate.

Reply:
Good suggestion. Addressed!